File: src/main/java/com/gitlab/daring/image/command/transform/RemoveLightCommand.java
package com.gitlab.daring.image.command.transform;

import com.gitlab.daring.image.command.BaseCommand;
import com.gitlab.daring.image.command.CommandEnv;
import com.gitlab.daring.image.command.parameter.IntParam;
import org.bytedeco.javacpp.opencv_core;

import static org.bytedeco.javacpp.opencv_imgproc.blur;

public class RemoveLightCommand extends BaseCommand {

    final IntParam blurRate = intParam(3, "0-255");

    public RemoveLightCommand(String... params) {
        super(params);
    }

    @Override
    public void execute(CommandEnv env) {
        opencv_core.Mat src = env.mat;
        opencv_core.Mat lightMap = new opencv_core.Mat(src.rows(), src.cols(), src.type());
        int br = blurRate.v;
        blur(src, lightMap, new opencv_core.Size(src.cols() / br, src.rows() / br));
        final opencv_core.Mat res = new opencv_core.Mat();
        opencv_core.subtract(src, lightMap, res);
        env.mat = res;
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/transform/RemoveLightCommand.kt
package com.gitlab.daring.image.command.transform

import com.gitlab.daring.image.command.CommandEnv
import com.gitlab.daring.image.command.KBaseCommand
import com.gitlab.daring.image.opencv.height
import com.gitlab.daring.image.opencv.width
import org.bytedeco.javacpp.opencv_core.*
import org.bytedeco.javacpp.opencv_imgproc.blur

internal class RemoveLightCommand(vararg params: String) : KBaseCommand(*params) {

    val blurRate = intParam(3, "0-255")

    val lightMap = Mat()
    val rm = Mat()

    override fun execute(env: CommandEnv) {
        val src = env.mat
        val br = blurRate.v
        blur(src, lightMap, Size(src.width / br, src.height / br))
        subtract(src, lightMap, rm)
        env.mat = rm
    }

}